### PR TITLE
[1/x]: Make Float8Linear support dynamic scaling

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #300
* #299
* #298
* #297
* #296
* #294
* #293
* #291
* __->__ #290

Summary:

At a high level, we need to make dynamic vs delayed scaling configurable
separately for activations, weights and gradients. The way I am
approaching this is as follows:
* PR 1 (this PR): add basic support for dynamic scaling, configurable by tensor, to `Float8Linear`
* PRs 2..n: one by one, add features implemented in `Float8DynamicLinear` to `Float8Linear`, as necessary
* last PR: delete `Float8DynamicLinear`

Test Plan:

```
./test/test_everything.sh
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D59305792](https://our.internmc.facebook.com/intern/diff/D59305792)

### CODE REVIEW

File: test/test_base.py [hunk 1]
```diff
@@ -208,14 +239,43 @@ def test_linear_nobias(
                     f"CUDA capability {torch.cuda.get_device_capability()} < (9.0)"
                 )
                 pytest.skip()
+        if linear_type is LinearType.DYNAMIC:
+            # Only test one combination of scaling types, as they are a no-op
```
Comment (vkuzo) [inline test/test_base.py:243]: this can be reusable but not worth it since it will be deleted further in the stack

File: float8_experimental/float8_linear.py [hunk 2]
```diff
@@ -326,11 +390,29 @@ def forward(self, x):
         if self.bias is not None:
             y = y + self.bias.to(y.dtype)
 
-        self.float8_post_forward()
+        if self.has_any_delayed_scaling:
+            self.float8_post_forward()
         return y
 
+    def extra_repr(self):
```
Comment (drisspg) [inline float8_experimental/float8_linear.py:397]: should we override the default repr to print this extra repr?

File: float8_experimental/float8_linear_utils.py [hunk 3]
```diff
@@ -30,35 +30,52 @@ class LinearType(Enum):
...
-def linear_requires_sync(linear_type: LinearType):
+def linear_requires_sync(
+    linear_type: LinearType,
+    scaling_type_x: TensorScalingType = TensorScalingType.DELAYED,
+    scaling_type_w: TensorScalingType = TensorScalingType.DELAYED,
+    scaling_type_dL_dY: TensorScalingType = TensorScalingType.DELAYED,
+):
     """Returns whether the given linear_type requires sync before forward."""
-    return linear_type in REQUIRES_SYNC
+    return linear_type is LinearType.DELAYED and any(
```
Comment (drisspg) [inline float8_experimental/float8_linear_utils.py:72]: should this be or?

File: float8_experimental/float8_linear_utils.py [hunk 4]
```diff
@@ -30,35 +30,52 @@ class LinearType(Enum):
...
-        LinearType.DYNAMIC: Float8DynamicLinear,
-    }
-    if linear_type not in LINEAR_TYPE_MAP:
-        raise ValueError(f"linear_type must be one of {LINEAR_TYPE_MAP.keys()}")
-    return LINEAR_TYPE_MAP[linear_type].from_float(
-        copy.deepcopy(linear_ref),
-        emulate=emulate,
-    )
+    if linear_type is LinearType.DYNAMIC:
+        return Float8DynamicLinear.from_float(
```
Comment (drisspg) [inline float8_experimental/float8_linear_utils.py:51]: should we assert that non of the scaling types are True?

File: float8_experimental/float8_linear.py (hunk 2 above)
Comment (drisspg) [reply float8_experimental/float8_linear.py:397]: ohh nvm from the tests it looks this is already done?

File: float8_experimental/float8_linear_utils.py [hunk 5]
```diff
@@ -30,35 +30,52 @@ class LinearType(Enum):
...
+            scaling_type_x=scaling_type_x,
+            scaling_type_w=scaling_type_w,
+            scaling_type_dL_dY=scaling_type_dL_dY,
+        )
 
 
-def linear_requires_sync(linear_type: LinearType):
+def linear_requires_sync(
+    linear_type: LinearType,
+    scaling_type_x: TensorScalingType = TensorScalingType.DELAYED,
```
Comment (drisspg) [inline float8_experimental/float8_linear_utils.py:67]: Nit: we should probably remove the defaults right?

File: test/test_base.py (hunk 1 above)
Comment (drisspg) [reply test/test_base.py:242]: we do use it in a lot of places lol but thats fine

**Review by drisspg — APPROVED** [review]
Seems good to me

File: float8_experimental/float8_linear.py (hunk 2 above)
Comment (vkuzo) [reply float8_experimental/float8_linear.py:397]: yes this is defined on `torch.nn.Module` for exactly this use case

File: float8_experimental/float8_linear_utils.py (hunk 5 above)
Comment (vkuzo) [reply float8_experimental/float8_linear_utils.py:67]: hmm, long term probably yes.  I'm taking the approach of tackling the "what's the default" recipe separately and not changing default behavior for now.

File: float8_experimental/float8_linear_utils.py (hunk 3 above)
Comment (vkuzo) [reply float8_experimental/float8_linear_utils.py:72]: since `Float8DynamicLinear` does not support `TensorScalingType`, I think `and` is right?

File: test/test_base.py (hunk 1 above)
Comment (vkuzo) [reply test/test_base.py:242]: I tried again just now to make it reusable and it was about the same LOC...:(

File: float8_experimental/float8_linear_utils.py (hunk 3 above)
Comment (drisspg) [reply float8_experimental/float8_linear_utils.py:72]: you are right 👍

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59305792).
